File: src/main/java/cz/encircled/eplayer/core/OnDemandFolderScanner.java
package cz.encircled.eplayer.core;

import cz.encircled.eplayer.model.MediaFile;
import cz.encircled.eplayer.model.PlayableMedia;
import cz.encircled.eplayer.service.FolderScanService;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * @author Encircled on 20/09/2014.
 */
public class OnDemandFolderScanner implements FolderScanService {

    private static final List<String> SUPPORTED_FORMATS = Arrays.asList("avi", "mkv", "mp3", "mp4", "flv", "wav", "wmv", "mov");

    private final ApplicationCore core;

    public OnDemandFolderScanner(@NotNull ApplicationCore core) {
        this.core = core;
    }

    @Override
    @NotNull
    public List<PlayableMedia> getMediaInFolder(@NotNull String path) {
        List<MediaFile> mediaFiles = new ArrayList<>();
        /*IOUtil.getFilesInFolder(path).forEach(file -> {
            if (SUPPORTED_FORMATS.contains(FilenameUtils.getExtension(file.getName()))) {
                PlayableMedia entry = core.getCacheService().getEntry(file.getPath());
                if (entry == null) {
                    entry = new MediaFile(file.getPath());
                }
                mediaFiles.add(entry);
            }
        }); TODO */
        return new ArrayList<>();
    }

}

File: src/main/java/cz/encircled/eplayer/core/OnDemandFolderScanner.kt
package cz.encircled.eplayer.core

import cz.encircled.eplayer.model.MediaSeries
import cz.encircled.eplayer.model.PlayableMedia
import cz.encircled.eplayer.service.CancelableExecution
import cz.encircled.eplayer.service.FolderScanService
import cz.encircled.eplayer.util.TimeMeasure
import java.io.File

/**
 * @author Encircled on 20/09/2014.
 */
class OnDemandFolderScanner(private val core: ApplicationCore) : FolderScanService {

    private val supportedFormats = setOf("avi", "mkv", "mp3", "mp4", "flv", "wav", "wmv", "mov", "ts")

    override fun getMediaInFolder(path: String, callback: CancelableExecution<List<PlayableMedia>>) {
        val result = TimeMeasure.measure("OnDemandFolderScanner - $path") {
            val series = HashSet<String>()
            File(path).walk().maxDepth(3)
                .filter { it.isFile && supportedFormats.contains(it.extension.lowercase()) && it.canRead() }
                .mapNotNull { playableForFile(it, series) }
                .toList()
        }

        callback.invoke(result)
    }


    private fun playableForFile(it: File, series: MutableSet<String>): PlayableMedia? {
        val media = core.cacheService.createIfAbsent(it.path)
        return if (media is MediaSeries) {
            if (series.add(media.name)) media else null
        } else media
    }

}
